packages/sdk/src/protocols/base_protocol.ts

```diff
@@ -1,15 +1,206 @@
+    }
+    return true;
+  };
+
+  /**
+   * Starts an interval to maintain the peers list to `numPeersToUse`.
+   * @param interval The interval in milliseconds to maintain the peers.
+   */
+  private async startMaintainPeersInterval(interval: number): Promise<void> {
+    this.log.info("Starting maintain peers interval");
+    try {
+      await this.maintainPeers();
+      this.maintainPeersIntervalId = setInterval(() => {
+        this.maintainPeers().catch((error) => {
+          this.log.error("Error during maintain peers interval:", error);
+        });
+      }, interval);
+      this.log.info(
+        `Maintain peers interval started with interval ${interval}ms`
+      );
+    } catch (error) {
+      this.log.error("Error starting maintain peers interval:", error);
+      throw error;
+    }
+  }
+
+  /**
+   * Maintains the peers list to `numPeersToUse`.
+   */
+  private async maintainPeers(): Promise<boolean> {
+    if (this.maintainPeersLock) {
+      this.log.info("Maintain peers already in progress, skipping");
+      return false;
+    }
+
+    this.maintainPeersLock = true;
+    this.log.info(`Maintaining peers, current count: ${this.peers.length}`);
+    try {
+      const numPeersToAdd = this.numPeersToUse - this.peers.length;
+      if (numPeersToAdd > 0) {
+        await this.findAndAddPeers(numPeersToAdd);
+      }
+      this.log.info(
+        `Peer maintenance completed, current count: ${this.peers.length}`
+      );
+    } catch (error) {
+      this.log.error("Error maintaining peers:", error);
+      throw error;
+    } finally {
+      this.maintainPeersLock = false;
+    }
+    return true;
+  }
+
+  /**
+   * Finds and adds new peers to the peers list.
+   * @param numPeers The number of peers to find and add.
+   */
+  private async findAndAddPeers(numPeers: number = 1): Promise<void> {
+    this.log.info(`Finding and adding ${numPeers} new peers`);
+    try {
+      const additionalPeers = await this.findAdditionalPeers(numPeers);
+      this.peers = [...this.peers, ...additionalPeers];
+      this.log.info(
+        `Added ${additionalPeers.length} new peers, total peers: ${this.peers.length}`
+      );
+    } catch (error) {
+      this.log.error("Error finding and adding new peers:", error);
+      throw error;
+    }
+  }
+
+  /**
+   * Finds additional peers.
+   * Attempts to find peers without using bootstrap peers first,
+   * If no peers are found,
+   * tries with bootstrap peers.
+   * @param numPeers The number of peers to find.
+   */
+  private async findAdditionalPeers(numPeers: number): Promise<Peer[]> {
+    this.log.info(`Finding ${numPeers} additional peers`);
+    try {
+      let newPeers = await this.core.getPeers({
+        maxBootstrapPeers: 0,
+        numPeers: numPeers
+      });
+
+      if (newPeers.length === 0) {
+        this.log.warn("No new peers found, trying with bootstrap peers");
+        newPeers = await this.core.getPeers({
```

Comment:
can you help me understand here: in this branch we try to get `newPeers` but since there are no just peers (from prev attempt) - then we try to get, potentially, same amount of bootstrap peers? 

if so, it seems we need to set `maxBootstraPeers` to `numPeers`